Share more code between BlobResourceHandle and NetworkDataTaskBlob

cdumez opened the pull request:
#### 7b3d928b91bddb673246051e0b68b19ed8444293
<pre>
Share more code between BlobResourceHandle and NetworkDataTaskBlob
<a href="https://bugs.webkit.org/show_bug.cgi?id=301032">https://bugs.webkit.org/show_bug.cgi?id=301032</a>

Reviewed by Darin Adler.

* Source/WebCore/fileapi/AsyncFileStream.cpp:
(WebCore::AsyncFileStream::perform):
* Source/WebCore/platform/FileStreamClient.h:
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandleBase::BlobResourceHandleBase):
(WebCore::BlobResourceHandleBase::closeFileIfOpen):
(WebCore::BlobResourceHandleBase::clearAsyncStream):
(WebCore::BlobResourceHandleBase::blobData const):
(WebCore::BlobResourceHandleBase::syncStream const):
(WebCore::BlobResourceHandleBase::asyncStream const):
(WebCore::BlobResourceHandle::BlobResourceHandle):
(WebCore::BlobResourceHandle::cancel):
(WebCore::BlobResourceHandleBase::start):
(WebCore::BlobResourceHandleBase::doStart):
(WebCore::BlobResourceHandleBase::getSizeForNext):
(WebCore::BlobResourceHandleBase::didGetSize):
(WebCore::BlobResourceHandleBase::seek):
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandleBase::readAsync):
(WebCore::BlobResourceHandleBase::readDataAsync):
(WebCore::BlobResourceHandleBase::readFileAsync):
(WebCore::BlobResourceHandleBase::didOpen):
(WebCore::BlobResourceHandleBase::didRead):
(WebCore::BlobResourceHandleBase::consumeData):
(WebCore::BlobResourceHandle::shouldAbortDispatchDidReceiveResponse):
(WebCore::BlobResourceHandleBase::dispatchDidReceiveResponse):
(WebCore::BlobResourceHandle::didReceiveResponse):
(WebCore::BlobResourceHandle::didReceiveData):
(WebCore::BlobResourceHandle::didFail):
(WebCore::BlobResourceHandle::didFinish):
(): Deleted.
(WebCore::BlobResourceHandle::start): Deleted.
(WebCore::BlobResourceHandle::doStart): Deleted.
(WebCore::BlobResourceHandle::getSizeForNext): Deleted.
(WebCore::BlobResourceHandle::didGetSize): Deleted.
(WebCore::BlobResourceHandle::readAsync): Deleted.
(WebCore::BlobResourceHandle::readDataAsync): Deleted.
(WebCore::BlobResourceHandle::readFileAsync): Deleted.
(WebCore::BlobResourceHandle::didOpen): Deleted.
(WebCore::BlobResourceHandle::didRead): Deleted.
(WebCore::BlobResourceHandle::consumeData): Deleted.
(WebCore::BlobResourceHandle::failed): Deleted.
(WebCore::BlobResourceHandle::notifyResponse): Deleted.
(WebCore::BlobResourceHandle::notifyResponseOnSuccess): Deleted.
(WebCore::BlobResourceHandle::notifyReceiveData): Deleted.
(WebCore::BlobResourceHandle::notifyFail): Deleted.
(WebCore::BlobResourceHandle::notifyFinish): Deleted.
* Source/WebCore/platform/network/BlobResourceHandle.h:
(WebCore::BlobResourceHandleBase::isFileOpen const):
(WebCore::BlobResourceHandleBase::setIsFileOpen):
(WebCore::BlobResourceHandleBase::async const):
(WebCore::BlobResourceHandleBase::totalSize):
(WebCore::BlobResourceHandleBase::totalRemainingSize const):
(WebCore::BlobResourceHandleBase::currentItemReadSize const):
(WebCore::BlobResourceHandleBase::setCurrentItemReadSize):
(WebCore::BlobResourceHandleBase::decrementTotalRemainingSizeBy):
(WebCore::BlobResourceHandleBase::readItemCount const):
(WebCore::BlobResourceHandleBase::incrementReadItemCount):
(WebCore::BlobResourceHandleBase::lengthOfItemBeingRead const):
(WebCore::BlobResourceHandleBase::buffer):
(WebCore::BlobResourceHandleBase::buffer const):
(WebCore::BlobResourceHandleBase::shouldAbortDispatchDidReceiveResponse):
(WebCore::BlobResourceHandleBase::clearStream):
* Source/WebCore/platform/network/ResourceHandle.cpp:
(WebCore::ResourceHandle::firstRequest const):
* Source/WebCore/platform/network/ResourceHandle.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::blobDataFrom):
(WebKit::NetworkDataTaskBlob::NetworkDataTaskBlob):
(WebKit::NetworkDataTaskBlob::clearStream):
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::cancel):
(WebKit::NetworkDataTaskBlob::erroredOrAborted const):
(WebKit::NetworkDataTaskBlob::didReceiveResponse):
(WebKit::NetworkDataTaskBlob::didReceiveData):
(WebKit::NetworkDataTaskBlob::download):
(WebKit::NetworkDataTaskBlob::writeDownload):
(WebKit::NetworkDataTaskBlob::getSizeForNext): Deleted.
(WebKit::NetworkDataTaskBlob::didGetSize): Deleted.
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse): Deleted.
(WebKit::NetworkDataTaskBlob::read): Deleted.
(WebKit::NetworkDataTaskBlob::readData): Deleted.
(WebKit::NetworkDataTaskBlob::readFile): Deleted.
(WebKit::NetworkDataTaskBlob::didOpen): Deleted.
(WebKit::NetworkDataTaskBlob::didRead): Deleted.
(WebKit::NetworkDataTaskBlob::consumeData): Deleted.
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:

Canonical link: <a href="https://commits.webkit.org/301798@main">https://commits.webkit.org/301798@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3abd59ca2a5ad3e1646f75ec795f406ff8e378f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37851 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78659 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d88d54ce-57cb-47cd-8e75-866add760e51) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96716 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64760 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d335bc0f-3dac-4222-bc48-7db1506ba857) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130045 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37902 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113818 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77226 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fcc30fd4-c363-4238-874c-afa617757539) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36772 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31943 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77491 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107769 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32300 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136625 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53752 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105236 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110174 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104925 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26757 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50452 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28882 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51301 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53684 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59557 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52922 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56255 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54680 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -427,67 +452,64 @@ int BlobResourceHandle::readFileSync(const BlobDataItem& item, std::span<uint8_t
+    while (m_totalRemainingSize && m_readItemCount < m_blobData->items().size()) {
+        const BlobDataItem& item = m_blobData->items().at(m_readItemCount);
+        switch (item.type()) {
+        case BlobDataItem::Type::Data:
+            if (!readDataAsync(item))
+                return; // error occurred
+            break;
+        case BlobDataItem::Type::File:
+            readFileAsync(item);
+            return;
```
Why do we `break` and continue looping for data blobs, but not for files? That seems to be new in the refactored code, couldn’t find that difference in the old code.

cdumez replied:
I will look into it to try and understand which behavior is correct here but I merely moved this code from [NetworkDataTaskBlob.cpp](https://github.com/WebKit/WebKit/pull/52608/files#diff-876270b7d298c66ef72f2ed7b3f2f6b5c4246472a6238328aa7326775981b46e).

cdumez replied:
As I said, I copied this logic from WebKit 2. The logic seems correct to me. When we encounter a File, we call `readFileAsync()`, which will use the `AsyncFileStream` to read the file asynchronously. This will end up calling `didRead()`, which after consuming the data will call `readAsync()` again:
```
void BlobResourceHandleBase::didRead(int bytesRead)
{
    // ...
    if (consumeData(m_buffer.subspan(0, bytesRead)))
        readAsync();
}
```

so then we'll resume the looping over the items.

darinadler commented on Source/WebCore/platform/FileStreamClient.h:
```diff
@@ -28,12 +28,13 @@
-#define FileStreamClient_h
+#include <wtf/AbstractRefCountedAndCanMakeWeakPtr.h>
+
+#pragma once
```
```suggestion
#pragma once

#include <wtf/AbstractRefCountedAndCanMakeWeakPtr.h>
```

cdumez replied:
Indeed, will fix.

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.h:
```diff
@@ -46,9 +47,7 @@ class ResourceHandleClient;
-// FIXME: Move even more logic from BlobResourceHandle / NetworkDataTaskBlob to
-// BlobResourceHandleBase and make data members private.
-class BlobResourceHandleBase {
+class BlobResourceHandleBase : public FileStreamClient {
```
Could `FileStreamClient` be a `private` or `protected` base class instead of `public`?

cdumez replied:
No, it won't build unless `public`:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.2.Internal.sdk/usr/include/c++/v1/__memory/unique_ptr.h:767:34: error: cannot cast 'WebCore::BlobResourceHandleBase' to its protected base class 'FileStreamClient'
/Volumes/Work/WebKit/OpenSource/Source/WebCore/platform/network/BlobResourceHandle.cpp:135:20: note: in instantiation of function template specialization 'WTF::makeUnique<WebCore::AsyncFileStream, WebCore::BlobResourceHandleBase &>' requested here
  135 |         m_stream = makeUnique<AsyncFileStream>(*this);
```

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.h:
```diff
@@ -58,77 +57,111 @@ class BlobResourceHandleBase {
     RefPtr<BlobData> m_blobData;
-    bool m_isRangeRequest { false };
+    std::unique_ptr<FileStream> m_syncStream; // For synchronous loading.
+    std::unique_ptr<AsyncFileStream> m_asyncStream; // For asynchronous loading.
```
Also, do we need a separate `m_async` boolean? Can we use nullity of `m_syncStream` instead?

Consider using a `Variant` here since it’s not a legal state for both to be non-null?

cdumez replied:
Yes, I can get rid of the m_async data member and use a WTF::Variant to replace those 2 members. I'll make this change before landing.

darinadler reviewed (approved):
Looks good to me, and very happy to see reduction in copied and pasted code. It’s just complex enough that we presumably need to rely on our testing coverage, not just looking it over.

cdumez committed:
Share more code between BlobResourceHandle and NetworkDataTaskBlob
https://bugs.webkit.org/show_bug.cgi?id=301032

Reviewed by Darin Adler.

* Source/WebCore/fileapi/AsyncFileStream.cpp:
(WebCore::AsyncFileStream::perform):
* Source/WebCore/platform/FileStreamClient.h:
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandleBase::BlobResourceHandleBase):
(WebCore::BlobResourceHandleBase::closeFileIfOpen):
(WebCore::BlobResourceHandleBase::clearAsyncStream):
(WebCore::BlobResourceHandleBase::blobData const):
(WebCore::BlobResourceHandleBase::syncStream const):
(WebCore::BlobResourceHandleBase::asyncStream const):
(WebCore::BlobResourceHandle::BlobResourceHandle):
(WebCore::BlobResourceHandle::cancel):
(WebCore::BlobResourceHandleBase::start):
(WebCore::BlobResourceHandleBase::doStart):
(WebCore::BlobResourceHandleBase::getSizeForNext):
(WebCore::BlobResourceHandleBase::didGetSize):
(WebCore::BlobResourceHandleBase::seek):
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandleBase::readAsync):
(WebCore::BlobResourceHandleBase::readDataAsync):
(WebCore::BlobResourceHandleBase::readFileAsync):
(WebCore::BlobResourceHandleBase::didOpen):
(WebCore::BlobResourceHandleBase::didRead):
(WebCore::BlobResourceHandleBase::consumeData):
(WebCore::BlobResourceHandle::shouldAbortDispatchDidReceiveResponse):
(WebCore::BlobResourceHandleBase::dispatchDidReceiveResponse):
(WebCore::BlobResourceHandle::didReceiveResponse):
(WebCore::BlobResourceHandle::didReceiveData):
(WebCore::BlobResourceHandle::didFail):
(WebCore::BlobResourceHandle::didFinish):
(): Deleted.
(WebCore::BlobResourceHandle::start): Deleted.
(WebCore::BlobResourceHandle::doStart): Deleted.
(WebCore::BlobResourceHandle::getSizeForNext): Deleted.
(WebCore::BlobResourceHandle::didGetSize): Deleted.
(WebCore::BlobResourceHandle::readAsync): Deleted.
(WebCore::BlobResourceHandle::readDataAsync): Deleted.
(WebCore::BlobResourceHandle::readFileAsync): Deleted.
(WebCore::BlobResourceHandle::didOpen): Deleted.
(WebCore::BlobResourceHandle::didRead): Deleted.
(WebCore::BlobResourceHandle::consumeData): Deleted.
(WebCore::BlobResourceHandle::failed): Deleted.
(WebCore::BlobResourceHandle::notifyResponse): Deleted.
(WebCore::BlobResourceHandle::notifyResponseOnSuccess): Deleted.
(WebCore::BlobResourceHandle::notifyReceiveData): Deleted.
(WebCore::BlobResourceHandle::notifyFail): Deleted.
(WebCore::BlobResourceHandle::notifyFinish): Deleted.
* Source/WebCore/platform/network/BlobResourceHandle.h:
(WebCore::BlobResourceHandleBase::isFileOpen const):
(WebCore::BlobResourceHandleBase::setIsFileOpen):
(WebCore::BlobResourceHandleBase::async const):
(WebCore::BlobResourceHandleBase::totalSize):
(WebCore::BlobResourceHandleBase::totalRemainingSize const):
(WebCore::BlobResourceHandleBase::currentItemReadSize const):
(WebCore::BlobResourceHandleBase::setCurrentItemReadSize):
(WebCore::BlobResourceHandleBase::decrementTotalRemainingSizeBy):
(WebCore::BlobResourceHandleBase::readItemCount const):
(WebCore::BlobResourceHandleBase::incrementReadItemCount):
(WebCore::BlobResourceHandleBase::lengthOfItemBeingRead const):
(WebCore::BlobResourceHandleBase::buffer):
(WebCore::BlobResourceHandleBase::buffer const):
(WebCore::BlobResourceHandleBase::shouldAbortDispatchDidReceiveResponse):
(WebCore::BlobResourceHandleBase::clearStream):
* Source/WebCore/platform/network/ResourceHandle.cpp:
(WebCore::ResourceHandle::firstRequest const):
* Source/WebCore/platform/network/ResourceHandle.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::blobDataFrom):
(WebKit::NetworkDataTaskBlob::NetworkDataTaskBlob):
(WebKit::NetworkDataTaskBlob::clearStream):
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::cancel):
(WebKit::NetworkDataTaskBlob::erroredOrAborted const):
(WebKit::NetworkDataTaskBlob::didReceiveResponse):
(WebKit::NetworkDataTaskBlob::didReceiveData):
(WebKit::NetworkDataTaskBlob::download):
(WebKit::NetworkDataTaskBlob::writeDownload):
(WebKit::NetworkDataTaskBlob::getSizeForNext): Deleted.
(WebKit::NetworkDataTaskBlob::didGetSize): Deleted.
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse): Deleted.
(WebKit::NetworkDataTaskBlob::read): Deleted.
(WebKit::NetworkDataTaskBlob::readData): Deleted.
(WebKit::NetworkDataTaskBlob::readFile): Deleted.
(WebKit::NetworkDataTaskBlob::didOpen): Deleted.
(WebKit::NetworkDataTaskBlob::didRead): Deleted.
(WebKit::NetworkDataTaskBlob::consumeData): Deleted.
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:

Canonical link: https://commits.webkit.org/301798@main
